### PR TITLE
Binary Protocol + Protocol Organization (step 1)

### PR DESCRIPTION
**Pull request by djrtwo**
I think we should get it merged in as is for now. Will allow Vlad to generate some binary gifs if he wants.

Stuff:
- your binary protocol
- add Protocol class that defines the components of a protocol (pass around as necessary)
- can specify protocol to casper.py via `--protocol` (defaults to blockchain) and must specify when making an experiment ("protocol")
- added some tests
- minimum separation in testing (much to still do)

NOTE: This is pointed at master. This includes your binary stuff but I already merged in master to here and handled a bunch of conflicts. I recommend merging this into master when ready and closing the other Binary PR.

### CODE REVIEW

**Comment by naterush** [comment]
Starting review now.

File: casper/binary/binary_view.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+from casper.binary.bet import Bet
+import casper.binary.binary_estimator as estimator
+
+import random as r
+
+
+class BinaryView(AbstractView):
+    """A view class that also keeps track of a last_finalized_block and children"""
+    def __init__(self, messages=None):
+        super().__init__()
```
Comment (naterush) [inline casper/binary/binary_view.py]: Don't add messages here through super, but down in BlockchainView. We should pick one.

File: casper/network.py [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+from casper.blockchain.blockchain_protocol import BlockchainProtocol
 
 
 class Network:
     """Simulates a network that allows for message passing between validators."""
-    def __init__(self, validator_set, view_class=BlockchainView):
+    def __init__(self, validator_set, protocol=BlockchainProtocol):
         self.validator_set = validator_set
-        self.global_view = view_class()
+        self.global_view = protocol.View(set())
```
Comment (naterush) [inline casper/network.py]: pretty :)

File: casper/binary/binary_view.py (hunk 1 above)
Comment (djrtwo) [reply casper/binary/binary_view.py]: I say do super.

You have to define `add_messages` in child, but the parent will always do the "if none, and add_message" call

**Comment by naterush** [comment]
This looks good. Merging now.

**Pull request merged**